File: 16. 3Sum Closest/16.cs
using System.Data.Common;

public class Solution {
    public int ThreeSumClosest(int[] nums, int target) {
        Array.Sort(nums);
        int closestSum = nums[0] + nums[1] + nums[2];

        if (nums.Length == 3)
        {
            return closestSum;
        }

        for(int i = 0; i < nums.Length -2; i++)
        {
            if (i>0 && nums[i] == nums[i - 1])
            {
                continue;
            }

            int j = i + 1;
            int k = nums.Length - 1;

            while (j < k)
            {

                int sumNums = nums[i] + nums[j] + nums[k];

                if (sumNums == target)
                    {
                        return sumNums;
                    }
                
                if (Math.Abs(target - sumNums) < Math.Abs(target - closestSum))
                {
                    closestSum = sumNums;
                }

                if (sumNums > target)
                {
                    k--;
                } else if (sumNums < target)
                {
                    j++;
                }
            }
        }
        return closestSum;
    }
}

File: 16. 3Sum Closest/16.py
class Solution:
    def threeSumClosest(self, nums: list[int], target: int) -> int:
        nums.sort()
        closestSum = nums[0] + nums[1] + nums[2]

        if len(nums) == 3:
            return closestSum


        for i in range(len(nums)-2):
            j = i + 1
            k = len(nums) -1

            while j<k:
                sumNums = nums[i] + nums[j] + nums[k]

                if abs(target - sumNums) < abs(target - closestSum):
                    closestSum = sumNums
                if closestSum == target:
                    return closestSum
                
                if sumNums > target:
                    k -= 1

                elif sumNums < target:
                    j += 1
            
        return closestSum
        

S = Solution()
print(S.threeSumClosest([1,1,1,1], 3))

                
